models/MortgageDetail.go

```diff
@@ -23,8 +26,127 @@ func GetMortgageDetails(userID uint32) ([]MortgageQueryData, error) {
 
-	db.Raw("SELECT stockId AS stock_id, SUM(stockQuantity) AS stocks_in_bank FROM Transactions WHERE userId = ? AND type = ? GROUP BY stockId", userID, "MortgageTransaction").Scan(&mortgageDetails)
+	db.Raw("SELECT stockId AS stock_id, stocksInBank AS stocks_in_bank, mortgagePrice AS mortgage_price  FROM MortgageDetails WHERE userId = ?", userID).Scan(&mortgageDetails)
 
 	l.Infof("Successfully fetched mortgageDetails for userID : %v", userID)
 	return mortgageDetails, nil
 }
+
+func (m *MortgageQueryData) ToProto() *models_pb.MortgageDetail {
+	return &models_pb.MortgageDetail{
+		StockId:       m.StockID,
+		StocksInBank:  m.StocksInBank,
+		MortgagePrice: m.MortgagePrice,
+	}
+}
+
+// mortgageStocksAction returns total transaction amount while mortgaging
+func mortgageStocksAction(user *User, stockID uint32, stockQuantity int64, mortgagePrice uint64, tx *gorm.DB) (int64, error) {
+	var l = logger.WithFields(logrus.Fields{
+		"method":              "mortgageStocksAction",
+		"param_userId":        user.Id,
+		"param_stockId":       stockID,
+		"param_stockQuantity": stockQuantity,
+		"param_mortgagePrice": mortgagePrice,
+	})
+
+	stockOwned, err := getSingleStockCount(user, stockID)
+	if err != nil {
+		l.Error(err)
+		return 0, err
+	}
+
+	if stockQuantity > stockOwned {
+		l.Errorf("Insufficient stocks to mortgage. Have %d, want %d", stockOwned, stockQuantity)
+		return 0, NotEnoughStocksError{}
+	}
+
+	stocksInBank, err := getStocksInBank(user.Id, stockID, mortgagePrice, tx)
+	if (err != nil && err != InvalidRetrievePriceError{} /* This type of error not applicable here */) {
+		l.Error(err)
+		return 0, err
+	}
+
+	if stocksInBank == 0 {
+		sql := "INSERT into MortgageDetails (userId, stockId, stocksInBank, mortgagePrice) VALUES (?, ?, ?, ?)"
+		err = tx.Exec(sql, user.Id, stockID, stockQuantity, mortgagePrice).Error
+	} else {
+		sql := "UPDATE MortgageDetails SET stocksInBank=? WHERE userId=? AND stockId=? AND mortgagePrice=?"
+		err = tx.Exec(sql, stocksInBank+stockQuantity, user.Id, stockID, mortgagePrice).Error
+	}
+
+	if err != nil {
+		l.Error(err)
+		tx.Rollback()
+		return 0, err
+	}
+
+	return int64(mortgagePrice) * stockQuantity * MORTGAGE_DEPOSIT_RATE / 100, nil
+}
+
+// retrieveStocksAction returns total transaction amount and stock quantity which could be retrieved
+func retrieveStocksAction(userID, stockID uint32, stockQuantity int64, userCash, retrievePrice uint64, tx *gorm.DB) (int64, error) {
+	var l = logger.WithFields(logrus.Fields{
+		"method":              "retrieveStocksAction",
+		"param_userId":        userID,
+		"param_stockId":       stockID,
+		"param_stockQuantity": stockQuantity,
+		"param_retrievePrice": retrievePrice,
+	})
+
+	l.Debugf("Retrieving stocks in action")
+
+	stocksInBank, err := getStocksInBank(userID, stockID, retrievePrice, tx)
+	if err != nil {
+		l.Infof("retrieveStocksActions failed for userId = %d, stockId = %d amount = %d while retrieving mortgaged stocks", userID, stockID, stockQuantity)
```

Comment:
Remove this log. `l.Error(...)` will print the same information. userID, stockID, stockQty etc will be printed because we've created `l` `WithFields` to print those for every statement it prints.